### ISSUE TITLE
tcp_server.c doesn't compile under GCC + socket fixes

### ISSUE THREAD

**Issue by shareefj** [lowRISC/opentitan#12395]
Hi,

I've been trying to use some of the DPI modules you have under hw/dv/dpi, especially reusing the TCP server, but it doesn't seem to compile using GCC.  It's quite possible I'm missing some switches but a quick search tells me this might be the difference between C++ and C and I think you probably only use it on OpenTitan under your C++ Verilator toplevel.

I'm using GCC 9.3 and had to make the following changes: https://github.com/idex-biometrics/opentitan/commit/cccadde51b6b398698431858a91d2dd85445a2e7

Also in that commit is a fix for the server not coping with dropped connections.  The assertion on L181 fails when tryaccept is called after the client disconnects: https://github.com/lowRISC/opentitan/blob/master/hw/dv/dpi/common/tcp_server/tcp_server.c#L181

**Comment by rswarbrick** [comment]
Hi there and thank you for the bug report.

I think you're right that we always compile this code with g++, which is why we won't have seen the build problems you're getting. Could you split this into two commits (separating "fix stupid build errors" and "handle dropped connections") and make a pull request?

**Comment by shareefj** [comment]
Yep, happy to do that.  Just have someone take a look at the logic behind the dropped connection change.  It works but it would be good to get a second pair of eyes to check that I haven't missed something.

**Comment by msfschaffner** [comment]
@moidx

**Comment by msfschaffner** [comment]
This could be addressed by #12397.

**Comment by rswarbrick** [comment]
Hmm, I think this is still pending(!). I've just looked at the proposed change. Adding the `struct` keyword is definitely right: we are currently writing C++ in a C file! :-) The handling of a client closing the connection is less "obviously right": we could also make the argument that we don't expect that to happen, and don't test it!

I'll push a PR in a second with the `struct` changes (and check that I can compile the result with GCC). Not so sure about the "socket fixes" part: we could reasonably say that they don't meet our expected use for the code. If @shareefj has different requirements, maybe the easiest thing is for him to fork it?

**Comment by rswarbrick** [comment]
About the pending C-compatibility changes: I've just looked a bit more closely (running `gcc -std=c99 -c hw/dv/dpi/common/tcp_server/tcp_server.c` to check).

The missing struct keyword was added in ef922c2615, back in September. I don't propose that we add `stdbool.h`: C99 has a `bool` keyword and our C code also has `//` comments, which aren't (strictly speaking) part of C89 either. Similarly, I'm not sure about replacing the `const int` with a `#define`: as far as I know, the `const int` approach is supported in C.

Amusingly, there *is* one thing that doesn't work with just C99: it turns out that `string.h` might not define `strdup()`. Apparently, it's only guaranteed to appear with C23! I'll push up a patch that defines the magic macro to include it. Otherwise, I propose that we close this issue.

Thank you very much, @shareefj, for noticing and reporting it!
